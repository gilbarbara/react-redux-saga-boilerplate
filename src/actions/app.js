// @flow
/**
 * @module Actions/App
 * @desc App Actions
 */

import uuid from 'uuid/v4';
import { createActions } from 'redux-actions';

import { ActionTypes } from 'constants/index';

export { goBack, go, push, replace } from 'modules/history';

export const { hideAlert, showAlert, switchMenu } = createActions({
  [ActionTypes.SWITCH_MENU]: (query: string) => ({ query }),
  [ActionTypes.HIDE_ALERT]: (id: string) => ({ id }),
  [ActionTypes.SHOW_ALERT]: (message: string, options: Object) => {
    const timeout = options.variant === 'danger' ? 0 : 5;

    return {
      id: options.id || uuid(),
      icon: options.icon,
      message,
      position: options.position || 'bottom-right',
      variant: options.variant || 'dark',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    };
  },
});
