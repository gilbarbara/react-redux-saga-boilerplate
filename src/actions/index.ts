export * from './app';
export * from './github';
export * from './user';
