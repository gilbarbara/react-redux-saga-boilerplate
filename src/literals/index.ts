import { keyMirror } from '@gilbarbara/helpers';

import { Status } from '~/types';

export const ActionTypes = keyMirror({
  ALERTS_HIDE: undefined,
  ALERTS_SHOW: undefined,
  GITHUB_GET_REPOS_REQUEST: undefined,
  GITHUB_GET_REPOS_SUCCESS: undefined,
  GITHUB_GET_REPOS_FAILURE: undefined,
  SET_APP_OPTIONS: undefined,
  USER_LOGIN_REQUEST: undefined,
  USER_LOGIN_SUCCESS: undefined,
  USER_LOGIN_FAILURE: undefined,
  USER_LOGOUT_REQUEST: undefined,
  USER_LOGOUT_SUCCESS: undefined,
  USER_LOGOUT_FAILURE: undefined,
});

export const STATUS: Status = {
  IDLE: 'idle',
  RUNNING: 'running',
  READY: 'ready',
  SUCCESS: 'success',
  ERROR: 'error',
};
