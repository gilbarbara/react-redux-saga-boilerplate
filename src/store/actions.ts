export { alertHide, alertShow } from './slices/alerts';
export { setAppOptions } from './slices/app';
export { getRepos, getReposFailure, getReposSuccess } from './slices/github';
export { login, loginSuccess, logOut, logOutSuccess } from './slices/user';
