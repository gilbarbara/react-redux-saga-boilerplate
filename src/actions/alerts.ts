import { ReactNode } from 'react';
import { uuid } from '@gilbarbara/helpers';
import { createAction } from '@reduxjs/toolkit';

import { actionPayload } from '~/modules/helpers';

import { ActionTypes } from '~/literals';

import { ShowAlertOptions } from '~/types';

export const hideAlert = createAction<string>(ActionTypes.ALERTS_HIDE);

export const showAlert = createAction(
  ActionTypes.ALERTS_SHOW,
  (message: ReactNode, options: ShowAlertOptions) => {
    const timeout = options.type === 'error' ? 0 : 5;

    return actionPayload({
      id: options.id || uuid(),
      icon: options.icon || 'info-o',
      message,
      position: options.position || 'bottom-right',
      type: options.type || 'neutral',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    });
  },
);
