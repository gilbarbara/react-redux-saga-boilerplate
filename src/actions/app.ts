/**
 * @module Actions/App
 * @desc App Actions
 */
import React from 'react';
import { nanoid } from 'nanoid';

import { createAction } from 'modules/helpers';

import { ActionTypes } from 'literals';

import { ShowAlertOptions } from 'types';

export { go, goBack, push, replace } from 'modules/history';

export const hideAlert = createAction(ActionTypes.HIDE_ALERT, (id: string) => ({ id }));

export const showAlert = createAction(
  ActionTypes.SHOW_ALERT,
  (message: React.ReactNode, options: ShowAlertOptions) => {
    const timeout = options.variant === 'danger' ? 0 : 5;

    return {
      id: options.id || nanoid(),
      icon: options.icon,
      message,
      position: options.position || 'bottom-right',
      variant: options.variant || 'dark',
      timeout: typeof options.timeout === 'number' ? options.timeout : timeout,
    };
  },
);
