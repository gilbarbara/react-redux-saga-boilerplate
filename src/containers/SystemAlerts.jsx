import React from 'react';
import PropTypes from 'prop-types';
import { connect } from 'react-redux';
import styled from 'styled-components';
import { utils } from 'styled-minimal';
import { hideAlert } from 'actions/index';

import Transition from 'components/Transition/index';
import Alert from 'components/Alert';

const Base = styled.div`
  position: fixed;
  z-index: 1000;

  > div {
    > * + * {
      margin-top: ${utils.spacer(3)};
    }
  }
`;

const TopLeft = styled(Base)`
  left: ${utils.spacer(3)};
  top: ${utils.spacer(3)};
  width: 26rem;

  /* stylelint-disable */
  ${utils.responsive({
    md: `
      width: 32rem;
    `,
  })};
  /* stylelint-enable */
`;

const TopRight = styled(Base)`
  right: ${utils.spacer(3)};
  top: ${utils.spacer(3)};
  width: 26rem;

  /* stylelint-disable */
  ${utils.responsive({
    md: `
      width: 32rem;
    `,
  })};
  /* stylelint-enable */
`;

const BottomLeft = styled(Base)`
  bottom: ${utils.spacer(3)};
  left: ${utils.spacer(3)};
  width: 26rem;

  /* stylelint-disable */
  ${utils.responsive({
    md: `
      width: 32rem;
    `,
  })};
  /* stylelint-enable */
`;

const BottomRight = styled(Base)`
  bottom: ${utils.spacer(3)};
  right: ${utils.spacer(3)};
  width: 26rem;

  /* stylelint-disable */
  ${utils.responsive({
    md: `
      width: 32rem;
    `,
  })};
  /* stylelint-enable */
`;

const SystemAlertsWrapper = styled.div`
  pointer-events: none;
  position: fixed;
  z-index: 1000;
`;

export class SystemAlerts extends React.PureComponent {
  constructor(props) {
    super(props);

    this.timeouts = {};
  }

  static propTypes = {
    app: PropTypes.object.isRequired,
    dispatch: PropTypes.func.isRequired,
  };

  componentWillReceiveProps(nextProps) {
    const {
      app: { alerts: nextAlerts },
      dispatch,
    } = nextProps;

    /* istanbul ignore else */
    if (nextAlerts.length) {
      nextAlerts.forEach(d => {
        if (d.timeout && !this.timeouts[d.id]) {
          this.timeouts[d.id] = setTimeout(() => {
            dispatch(hideAlert(d.id));
          }, d.timeout * 1000);
        }
      });
    }
  }

  componentWillUnmount() {
    Object.keys(this.timeouts).forEach(d => {
      clearTimeout(this.timeouts[d]);
    });
  }

  handleClick = e => {
    e.preventDefault();
    const { dataset } = e.currentTarget;
    const { dispatch } = this.props;

    dispatch(hideAlert(dataset.id));
  };

  renderAlerts(position) {
    const { app } = this.props;
    const items = app.alerts.filter(d => d.position === position);

    if (!items.length) {
      return null;
    }

    return items.map(d => (
      <Alert
        key={d.id}
        id={d.id}
        icon={d.icon}
        handleClickClose={this.handleClick}
        variant={d.variant}
      >
        {d.message}
      </Alert>
    ));
  }

  render() {
    return (
      <SystemAlertsWrapper key="SystemAlerts">
        <TopLeft>
          <Transition transition="slideDown">{this.renderAlerts('top-left')}</Transition>
        </TopLeft>
        <TopRight>
          <Transition transition="slideDown">{this.renderAlerts('top-right')}</Transition>
        </TopRight>
        <BottomLeft>
          <Transition transition="slideUp">{this.renderAlerts('bottom-left')}</Transition>
        </BottomLeft>
        <BottomRight>
          <Transition transition="slideUp">{this.renderAlerts('bottom-right')}</Transition>
        </BottomRight>
      </SystemAlertsWrapper>
    );
  }
}

/* istanbul ignore next */
function mapStateToProps(state) {
  return { app: state.app };
}

export default connect(mapStateToProps)(SystemAlerts);
