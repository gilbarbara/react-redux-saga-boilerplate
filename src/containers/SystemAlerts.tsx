import React, { useCallback, useEffect, useRef } from 'react';
import { useDispatch } from 'react-redux';
import { useUnmount } from 'react-use';
import styled from '@emotion/styled';
import { responsive } from '@gilbarbara/components';

import { useAppSelector } from '~/modules/hooks';
import theme from '~/modules/theme';

import { hideAlert } from '~/actions';

import Alert from '~/components/Alert';
import Transition from '~/components/Transition';

const Base = styled.div`
  position: fixed;
  z-index: 1000;

  > div {
    > * + * {
      margin-top: ${theme.spacing.md};
    }
  }
`;

const TopLeft = styled(Base)`
  left: ${theme.spacing.md};
  top: ${theme.spacing.md};
  width: 260px;

  ${
    /* sc-custom '@media-query' */ responsive({
      md: {
        width: '32rem',
      },
    })
  };
`;

const TopRight = styled(Base)`
  right: ${theme.spacing.md};
  top: ${theme.spacing.md};
  width: 260px;

  ${
    /* sc-custom '@media-query' */ responsive({
      md: {
        width: '32rem',
      },
    })
  };
`;

const BottomLeft = styled(Base)`
  bottom: ${theme.spacing.md};
  left: ${theme.spacing.md};
  width: 260px;

  ${
    /* sc-custom '@media-query' */ responsive({
      md: {
        width: '32rem',
      },
    })
  };
`;

const BottomRight = styled(Base)`
  bottom: ${theme.spacing.md};
  right: ${theme.spacing.md};
  width: 260px;

  ${
    /* sc-custom '@media-query' */ responsive({
      md: {
        width: '32rem',
      },
    })
  };
`;

const SystemAlertsWrapper = styled.div`
  pointer-events: none;
  position: fixed;
  z-index: 1000;
`;

export default function SystemAlerts() {
  const dispatch = useDispatch();
  const alerts = useAppSelector(s => s.alerts.data);
  const timeouts: Record<string, any> = useRef({});

  useEffect(() => {
    const { current } = timeouts;

    if (alerts.length) {
      alerts.forEach(d => {
        if (d.timeout && !current[d.id]) {
          current[d.id] = setTimeout(() => {
            dispatch(hideAlert(d.id));
          }, d.timeout * 1000);
        }
      });
    }
  }, [alerts, dispatch]);

  useUnmount(() => {
    const { current } = timeouts;

    Object.keys(current).forEach(d => {
      clearTimeout(current[d]);
    });
  });

  const handleClick = useCallback(
    (event: React.MouseEvent<HTMLButtonElement>) => {
      event.preventDefault();
      const { id = '' } = event.currentTarget.dataset;

      dispatch(hideAlert(id));
    },
    [dispatch],
  );

  const renderAlerts = useCallback(
    (position: string) => {
      const items = alerts.filter(d => d.position === position);

      if (!items.length) {
        return null;
      }

      return items.map(d => (
        <Alert key={d.id} handleClickClose={handleClick} icon={d.icon} id={d.id} type={d.type}>
          {d.message}
        </Alert>
      ));
    },
    [alerts, handleClick],
  );

  return (
    <SystemAlertsWrapper key="SystemAlerts">
      <TopLeft>
        <Transition transition="slideRight">{renderAlerts('top-left')}</Transition>
      </TopLeft>
      <TopRight>
        <Transition transition="slideLeft">{renderAlerts('top-right')}</Transition>
      </TopRight>
      <BottomLeft>
        <Transition transition="slideRight">{renderAlerts('bottom-left')}</Transition>
      </BottomLeft>
      <BottomRight>
        <Transition transition="slideLeft">{renderAlerts('bottom-right')}</Transition>
      </BottomRight>
    </SystemAlertsWrapper>
  );
}
