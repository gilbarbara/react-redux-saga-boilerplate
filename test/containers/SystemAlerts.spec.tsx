import React from 'react';
import { config } from 'react-transition-group';
import { act, fireEvent, render, screen } from 'test-utils';

import { showAlert } from '~/actions';
import { ActionTypes } from '~/literals';

import SystemAlerts from '~/containers/SystemAlerts';

config.disabled = true;

vi.mock('~/components/Transition', () => ({
  default: ({ children }) => <div className="transition">{children}</div>,
}));

vi.useFakeTimers();

const mockDispatch = vi.fn();

describe('SystemAlerts', () => {
  it('should render all zones', () => {
    const { container } = render(<SystemAlerts />);

    expect(container).toMatchSnapshot();
  });

  it('should render an alert and hide itself after the timeout', async () => {
    render(<SystemAlerts />, {
      actions: [
        showAlert('Hello World', {
          position: 'top-left',
          type: 'success',
        }),
      ],
      mockDispatch,
    });

    expect(screen.getByRole('alert')).toBeInTheDocument();

    act(() => {
      vi.runOnlyPendingTimers();
    });

    expect(mockDispatch).toHaveBeenCalledWith({
      type: ActionTypes.ALERTS_HIDE,
      payload: '8cdee72f-28d4-4441-91f0-c61f6e3d9684',
    });

    expect(screen.queryByRole('alert')).not.toBeInTheDocument();
  });

  it('should render an alert without timeout and close it', () => {
    render(<SystemAlerts />, {
      actions: [
        showAlert('Hello Mars', {
          id: 'ABD13',
          position: 'bottom-right',
          type: 'neutral',
          timeout: 0,
        }),
      ],
      mockDispatch,
    });

    expect(screen.getByRole('alert')).toBeInTheDocument();

    fireEvent.click(screen.queryByRole('alert')?.querySelector('button')!);

    expect(screen.queryByRole('alert')).not.toBeInTheDocument();
  });
});
