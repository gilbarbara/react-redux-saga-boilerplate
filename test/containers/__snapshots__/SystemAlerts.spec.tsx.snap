// Vitest Snapshot v1, https://vitest.dev/guide/snapshot.html

exports[`SystemAlerts > should render all zones 1`] = `
<div>
  <div
    class="css-1ewh9kd"
  >
    <div
      class="css-vg32sl"
    >
      <div
        class="transition"
      />
    </div>
    <div
      class="css-1r5pw8k"
    >
      <div
        class="transition"
      />
    </div>
    <div
      class="css-50gqfe"
    >
      <div
        class="transition"
      />
    </div>
    <div
      class="css-ypj8vw"
    >
      <div
        class="transition"
      />
    </div>
  </div>
</div>
`;
